src/host/alloc_cache.jl

```diff
@@ -24,43 +24,51 @@ end
 function get_pool!(cache::AllocCache, pool::Symbol, uid::UInt64)
     pool = getproperty(cache, pool)
     uid_pool = get(pool, uid, nothing)
-    if uid_pool ≡ nothing
-        uid_pool = Base.@lock cache.lock pool[uid] = Any[]
+    if uid_pool === nothing
+        uid_pool = pool[uid] = DataRef[]
     end
     return uid_pool
 end
 
 function cached_alloc(f, key)
     cache = ALLOC_CACHE[]
     if cache === nothing
-        return f()::AbstractGPUArray
+        return f()::DataRef
     end
 
-    x = nothing
+    ref = nothing
     uid = hash(key)
 
-    busy_pool = get_pool!(cache, :busy, uid)
-    free_pool = get_pool!(cache, :free, uid)
-    isempty(free_pool) && (x = f()::AbstractGPUArray)
+    Base.@lock cache.lock begin
+        free_pool = get_pool!(cache, :free, uid)
+
+        if !isempty(free_pool)
+            ref = Base.@lock cache.lock pop!(free_pool)
+            @assert !ref.freed
```

Comment:
This essentially forbids user calling `CUDA.unsafe_free!(x)` on an array inside `GPUArrays.@cached`.

I imagined `GPUArrays.@cached` to be like a toggle, such that with and without it the code can be used without any modifications.
Even if putting aside things like users calling `unsafe_free!` directly I think it will introduce problems with things like `with_workspace`.

Why not cycle in a loop until we find a non-freed array as it was before?

Reply:
Ah, good catch.

> Why not cycle in a loop until we find a non-freed array as it was before?

Because only the reference is freed, not the underlying data. I should find a way to still reuse that (but we can't simply toggle `freed` because then the GC may free it again).

Reply:
I've tested on GaussianSplatting.jl and HiFiGAN.jl and it works fine.
Let me know what you think of this approach.

Reply:
Hmm, I'm not a fan of coupling DataRef to the cache (which to me conceptually is a lower-level abstraction). I'll try to come up with a different approach Monday, or else go with your suggested one.

Reply:
Hmm, couldn't find a way around this. I did simplify your approach though; I don't think we need to check the scoped value in the `DataRef` ctor (but just set the value when caching a ref), and with this added logic we should be able to get rid of the manual `retain`/`release` in order to keep the underlying data from getting freed. Thoughts?